src/lib/core/exception.h

```diff
@@ -231,14 +292,42 @@ class CryptoError: public Exception {
 	}
 
 	virtual void raise() { throw this; }
+
+	virtual struct error *dup() const
+	{
+		return dup_error<CryptoError>(this);
+	}
 };
 
 class RaftError: public Exception {
 public:
+	RaftError() : Exception(&type_RaftError, NULL, 0) {}
+
 	RaftError(const char *file, unsigned line, const char *format, ...);
 	virtual void raise() { throw this; }
+	virtual struct error *dup() const
+	{
+		return dup_error<RaftError>(this);
+	}
 };
 
+template <class ErrorT>
+struct error *
+Exception::dup_error(const struct error *to_dup)
```

Comment:
This function seems to be a copy constructor. In C++ there's already a special way to define a copy constructor. Please use C++ way in C++ code.

Reply:
This is not actually a copy constructor, because it duplicates only received error and does not copy any cause and effect errors. I'l modify this function to `Exception *dup_one_exception<ErrorT>(const Exception *e)`, to make it less confusing and more similar to the C++ way.

Reply:
I think such a copy constructor might confuse someone. It looks a bit messy if you have a copy constructor in a C++ module that only copies the `Exception` object without any relative nodes, but at the same time in the C module you have the `error_copy` function that copies the `Error` object with all its cause nodes. 
I am going to implenet a copy constructor, which makes a deep copy(with cause nodes) or throws an `OutOfMemory` exception if the memroy allocation fails. Copying without cause nodes can lead to a confusing error. If the `dup` C-function returns the same `OutOfMemory` exception twice you will break the error list while linking error nodes. I can`t copy this exception because I need to allocate memory for its copy.
Alternatively, I can use xmalloc for panic if memory allocation fails, but I want to handle errors by using exceptions (as the master branch code) to avoid different ways of memory allocation and error handling.

Reply:
Hi, I'm a little confused with exception handling in this function which is used in C code for calling virtual methods of Exception object by using its base error object. What do you think of this default catch block? @alyapunov 
```C++
static struct error *
exception_dup(const struct error *error, bool *failed)
{
	try {
		Exception *e = (Exception *) error;
		return e->dup();
	} catch (Exception *e) {
		*failed = true;
		return e;
	} catch (...) {
		/** Exceptions are not allowed in C. */
		panic("Unexpected exception while copying");
	}
}
```